### PR TITLE
Introduce Gulp (with TypeScript transpiling)

### LINKED ISSUE #21: Use ES6 imports and TypeScript
Let's add TypeScript to our project and use `import`/`export` instead of `require`/`module.exports`. 

I don't insist that you should transform all your JS files to TS files right now. Just try setting `allowJS: true` & `checkJs: false` and compile your JS files without type checking.

Some other hints about the initial configuration:
`esModuleInterop: true`
`target: es2018`

Note: we can use TypeScript for client scripts too, and enable the built-in SystemJS bundling.

### PR DESCRIPTION
**Pull request by wentwrong**
# New changes
- Gulp with TypeScript plugin; (resolve #21)
<s>- Entry point (`index.js`) moved to root directory of project _(i think it looks more naturally for node package - easier to find)_;</s>
- Launching mocha tests _(already transpiled)_ via gulp;

## Food for thought
- ESLint via gulp? (troubles with `--fix`-ing files);
- Better `gulpfile.js` organization? _(Some `parallel`-ization?)_
- <s>Different `tsconfig.json` for client and server, explicitly specified with path in `gulpfile.js`. Could we in gulp before transpiling phase **automatically** pick up settings from `tsconfig.json` files _(without explicit path defining)_?</s>
- SystemJS in our `.html` file is imported due to CDN. Replace with local lib?

### CODE REVIEW

File: src/client/public/js/script.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-class App {
+export class App {
```
Comment (AndreyBelym) [inline src/client/public/js/script.js]: Why do we need to move this file to the `/js` subdirectory? Also, I thought that we will convert it to TS when the time comes.

File: gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            reporter: 'list',
+            exit:     true
+        }));
+});
+
+gulp.task('clean', async () => {
+    await del(['dist']);
+});
+
+gulp.task('build-client', gulp.series('copy-html', 'transpile-client'));
```
Comment (AndreyBelym) [inline gulpfile.js]: Why can't we copy HTML, build client and server scripts with `gulp.parallel`?

File: gulpfile.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        .js.pipe(gulp.dest(`${paths.BUILD_DIR}/src/client/public/js`));
+});
+
+gulp.task('copy-html', () => {
+    return gulp.src([`${paths.CLIENT_DIR}/*.html`])
+        .pipe(gulp.dest(`${paths.BUILD_DIR}/${paths.CLIENT_DIR}`));
+});
+
+gulp.task('copy-fixtures', () => {
+    return gulp.src([`${paths.TESTS_DIR}/fixtures/*.json`])
```
Comment (AndreyBelym) [inline gulpfile.js]: Also, I'm not fond of copying test data after each build. Need to think about the project structure.

File: src/client/public/js/script.js (hunk 1 above)
Comment (wentwrong) [reply src/client/public/js/script.js]: The only reason that i did that was to separate `.html` from source files. In the future, that folder may contain multiple `.js` files _(e.g. utils for `debug`)_, if we already refused idea of bundling.

File: gulpfile.js (hunk 3 above)
Comment (wentwrong) [reply gulpfile.js]: Solution: use ts-node.

**Comment by wentwrong** [comment]
`src` subfolder in `dist` seems a bit extra. Isn't it?

**Comment by AndreyBelym** [comment]
IMHO we should try to stick with the following structure:
- `/src` - all TS/JS source files, including the client ones
- `/lib` - transpiled JS files
- `/resources` - root of resources (`/resources/html` for HTML, `/resources/css` for CSS, etc.)
- `/tests` - for tests

File: src/client/public/js/script.js (hunk 1 above)
Comment (AndreyBelym) [reply src/client/public/js/script.js]: We shouldn't mix HTML files and other resources with the source files. They should be stored in a separate `/resources` or `/assets` directory. We can just add an another `app.static` for scripts.

**Comment by wentwrong** [comment]
So, the project structure looks like this:

```
gh-canary
├── lib
|  ├── client
|  └── server
|     ├── controllers
|     ├── middleware
|     └── routes
|        └── api
|           └── v1
|              └── pulls
├── resources
|  └── html
├── src
|  ├── client
|  |  └── controllers
|  └── server
|     ├── controllers
|     ├── middleware
|     └── routes
|        └── api
|           └── v1
|              └── pulls
└── test
   ├── controllers
   ├── fixtures
   └── routes
      └── mock-github
         └── repos
```

**Pull request merged**